Src/Base/AMReX_FabArrayCommI.H

```diff
@@ -408,73 +436,120 @@ FabArray<FAB>::ParallelCopy (const FabArray<FAB>& src,
 #ifdef AMREX_USE_GPU
             if (Gpu::inLaunchRegion())
             {
-                PC_local_gpu(thecpc, src, SC, DC, NC, op);
+                PC_local_gpu(thecpc, src, pc_SC, pc_DC, pc_NC, op);
             }
             else
 #endif
             {
-                PC_local_cpu(thecpc, src, SC, DC, NC, op);
+                PC_local_cpu(thecpc, src, pc_SC, pc_DC, pc_NC, op);
             }
         }
 
-        if (N_rcvs > 0)
+        if (!last_iter)
         {
-            Vector<const CopyComTagsContainer*> recv_cctc(N_rcvs,nullptr);
-	    for (int k = 0; k < N_rcvs; ++k)
-	    {
-                if (recv_size[k] > 0)
-                {
-                    auto const& cctc = thecpc.m_RcvTags->at(recv_from[k]);
-                    recv_cctc[k] = &cctc;
-                }
-	    }
+            ParallelCopy_finish();
 
-            if (actual_n_rcvs > 0) {
-                Vector<MPI_Status> stats(N_rcvs);
-                ParallelDescriptor::Waitall(recv_reqs, stats);
-#ifdef AMREX_DEBUG
-                if (!CheckRcvStats(stats, recv_size, SeqNum))
-                {
-                    amrex::Abort("ParallelCopy failed with wrong message size");
-                }
-#endif
-            }
+            pc_SC += pc_NC;
+            pc_DC += pc_NC;
+        }
 
-            bool is_thread_safe = thecpc.m_threadsafe_rcv;
+        ipass     += pc_NC;
+        NCompLeft -= pc_NC;
+    }
 
-#ifdef AMREX_USE_GPU
-            if (Gpu::inLaunchRegion())
-            {
-                unpack_recv_buffer_gpu(*this, DC, NC, recv_data, recv_size, recv_cctc,
-                                       op, is_thread_safe);
-            }
-            else
-#endif
+    return;
+
+#endif /*BL_USE_MPI*/
+}
+
+template <class FAB>
+void
+FabArray<FAB>::ParallelCopy_finish ()
+{
+    BL_PROFILE("FabArray::ParallelCopy_finish()");
+
+    if (size() == 0 || pc_src->size() == 0) return;
+
+    if ((pc_src->boxArray().ixType().cellCentered() || pc_op == FabArrayBase::COPY) &&
```

Comment:
The test here is non-trivial.  So maybe we should add a boolean flag and set it in the `nowait` function for this shortcut test and others that allow an immediate return.